### PR TITLE
Add liquidlegions formulas.

### PR DESCRIPTION
**Pull request by jiayu-google**

### CODE REVIEW

File: src/simulator/halo_simulator.py [hunk 1]
```diff
@@ -213,6 +214,46 @@ def simulated_reach_by_spend(
         kplus_reaches = ReachPoint.frequencies_to_kplus_reaches(frequencies)
         return ReachPoint(impressions, kplus_reaches, spends)
 
+    def _liquid_legions_cardinality_estimate_variance(self, n: int) -> float:
+        """Variance of cardinality estimate by unnoised LiquidLegions.
```
Comment (matthewclegg) [inline src/simulator/halo_simulator.py:218]: Is there a reference that you can give for this formula?  Perhaps that would be helpful for people outside of our team.

File: src/simulator/halo_simulator.py [hunk 2]
```diff
@@ -213,6 +214,46 @@ def simulated_reach_by_spend(
...
+        a = self._params.liquid_legions.decay_rate
+        c = a * n / m / (1 - np.exp(-a))
+        variance = (
+            expi(-c) - expi(-c * np.exp(-a)) - expi(-2 * c) + expi(-2 * c * np.exp(-a))
+        )
+        variance *= a * n ** 2 / m / (np.exp(-c) - np.exp(-c * np.exp(-a))) ** 2
+        variance -= n
+        return variance
+
+    def _liquid_legions_num_active_regions(
```
Comment (matthewclegg) [inline src/simulator/halo_simulator.py:238]: I wonder if the name _liquid_legions_sample_num_active_regions would be more accurate?

**Review by matthewclegg — CHANGES_REQUESTED** [review]

File: src/simulator/halo_simulator.py (hunk 1 above)
Comment (jiayu-google) [reply src/simulator/halo_simulator.py:218]: It is included in the PoPETs paper which has been accepted and will hopefully be published soon.  No public reference off the top of my head.  Maybe we can add a doc to the WFA repo (let me first confirm if the paper is formally accepted).

File: src/simulator/halo_simulator.py (hunk 2 above)
Comment (jiayu-google) [reply src/simulator/halo_simulator.py:238]: I believe that num_active_registers is the most commonly used name within Google, and sample_num_active_regions or simply `sampled_regions` also make sense to me.  Now I prefer `num_active_registers`, but is open to other options.

File: src/simulator/halo_simulator.py [hunk 3]
```diff
@@ -213,6 +214,46 @@ def simulated_reach_by_spend(
...
+        self, n: int, random_generator: np.random.Generator = np.random.default_rng()
```
Comment (riemanli) [inline src/simulator/halo_simulator.py:239]: I know we have done this several times, but I always wonder why we need to have a different random generator when there is one in `self._param`.

File: src/simulator/halo_simulator.py (hunk 1 above)
Comment (jiayu-google) [reply src/simulator/halo_simulator.py:218]: Added a TODO here.

File: src/simulator/halo_simulator.py (hunk 3 above)
Comment (jiayu-google) [reply src/simulator/halo_simulator.py]: Thank you for pointing this out!  Revised following your practice in PR 56.

**Review by riemanli — APPROVED** [review]

**Review by matthewclegg — APPROVED** [review]

File: src/simulator/halo_simulator.py [hunk 4]
```diff
@@ -213,6 +214,50 @@ def simulated_reach_by_spend(
...
+          An observation of the number of active registers in the LiquidLegions.
+        """
+        m = self._params.liquid_legions.sketch_size
+        a = self._params.liquid_legions.decay_rate
+        register_probs = np.exp(-a * np.arange(m) / m)
+        register_probs /= sum(register_probs)
+        random_generator = np.random.default_rng(
+            seed=self._params.generator.integers(low=0, high=1e9)
+        )
+        return sum(random_generator.multinomial(n, register_probs) == 1)
```
Comment (riemanli) [inline src/simulator/halo_simulator.py:259]: Would it be simpler to directly use `self._params.generator` for `multinomial`?

```diff
- random_generator = np.random.default_rng(
-     seed=self._params.generator.integers(low=0, high=1e9)
- )
- return sum(random_generator.multinomial(n, register_probs) == 1)
+ return sum(self._params.generator.multinomial(n, register_probs) == 1)
```

**Comment by RonPinelliMRC** [comment]
Do @jiayu-google's most recent changes need internal review before MRC s/o?

**Review by riemanli — APPROVED** [review]

**Comment by riemanli** [comment]
@RonPinelliMRC, thanks for raising the concern. 
Yes, they do. The internal review is now done and ready for MRC s/o. 
Thank you!

**Review by RonPinelliMRC — APPROVED** [review]
MRC approving based on cursory review after confirming non-independent code review on an exception basis. Experimental code will require independent review prior to publication of or reliance on results generated.

**Pull request merged**